batch/sql/estimated-current.sql

```diff
@@ -1531,7 +1541,15 @@ BEGIN
            jobs.job_id = `job_parents`.job_id
       SET jobs.state = IF(jobs.n_pending_parents = 1, 'Ready', 'Pending'),
           jobs.n_pending_parents = jobs.n_pending_parents - 1,
-          jobs.cancelled = IF(new_state = 'Success', jobs.cancelled, 1)
+          jobs.n_succeeded_parents = jobs.n_succeeded_parents + (new_state = 'Success'),
+          jobs.cancelled = (CASE jobs.run_condition
+                WHEN 'always' THEN jobs.cancelled
+                WHEN 'all_succeeded' THEN IF(new_state = 'Success', jobs.cancelled, 1)
+                WHEN 'any_succeeded' THEN IF(COALESCE(jobs.n_pending_parents - 1, 0) = 0,
+                                             IF(COALESCE(jobs.n_succeeded_parents, 0) + (new_state = 'Success') > 0, jobs.cancelled, 1),
+                                             jobs.cancelled)
+                ELSE IF(new_state = 'Success', jobs.cancelled, 1)  # backwards compatibility
+                END)
```

Comment:
Sorry I'm still finding this quite hard to parse. Is the following equivalent? I find it easier to frame the answer to the question as "The job is cancelled if \<some condition is met\> OR the job was already cancelled". This way we never uncancel a job which I feel like is an important property.

```
CASE jobs.run_condition
WHEN 'always' THEN jobs.cancelled
WHEN 'all_succeeded' THEN new_state != 'Success' OR jobs.cancelled
WHEN 'any_succeeded' THEN (jobs.n_pending_parents - 1 = 0 AND jobs.n_succeeded_parents = 0 AND new_state != 'Success') OR jobs.cancelled
ELSE new_state != 'Success' or jobs.cancelled
END
```